File: src/civ_core/api/handlers/pdf_tools.py
"""pdf_tools handlers：PDF 合并/拆分 RPC 接口。

RPC 方法（前缀 "pdf_tools."）：
  pdf_tools.merge(inputs, output)
    -> {output: str, count: int}
  pdf_tools.split_per_page(input, output_dir, name_template="{stem}_p{n}.pdf")
    -> {written: [str], count: int}
  pdf_tools.split_by_ranges(input, output_dir, expr, name_template="{stem}_{start}-{end}.pdf")
    -> {written: [str], count: int}
  pdf_tools.inspect(paths)
    -> {files: [{path, pages, size_kb, error?}], total_pages: int}
       给前端「中间预览」用：列出每个 PDF 的页数 + 大小，单个失败不影响整体。
"""

from __future__ import annotations

from pathlib import Path

from civ_core.infra_io.pdf_io import (
    merge_pdfs,
    split_pdf_by_ranges,
    split_pdf_per_page,
)

__all__ = ["merge", "split_per_page", "split_by_ranges", "inspect"]


def merge(inputs: list[str], output: str) -> dict:
    """按 inputs 顺序合并为单个 PDF。"""
    out = merge_pdfs([Path(p) for p in inputs], Path(output))
    return {"output": str(out), "count": len(inputs)}


def split_per_page(
    input: str,  # noqa: A002 — RPC 接口键固定为 "input"
    output_dir: str,
    name_template: str = "{stem}_p{n}.pdf",
) -> dict:
    written = split_pdf_per_page(Path(input), Path(output_dir), name_template=name_template)
    return {"written": [str(p) for p in written], "count": len(written)}


def split_by_ranges(
    input: str,  # noqa: A002
    output_dir: str,
    expr: str,
    name_template: str = "{stem}_{start}-{end}.pdf",
) -> dict:
    written = split_pdf_by_ranges(Path(input), Path(output_dir), expr, name_template=name_template)
    return {"written": [str(p) for p in written], "count": len(written)}


def inspect(paths: list[str]) -> dict:
    """读每个 PDF 的页数 + 文件大小给前端预览（合并模式列表 / 拆分模式单文件信息）。

    单个文件失败（路径错 / 加密 / 损坏）→ files[i] 带 error 字段，不影响其他文件
    和合计逻辑。total_pages 只累加 pages 不为 None 的项。
    """
    from pypdf import PdfReader
    from pypdf.errors import PdfReadError

    files: list[dict] = []
    total = 0
    for p in paths:
        path = Path(p)
        item: dict = {"path": p}
        if not path.is_file():
            item["error"] = f"文件不存在：{p}"
            files.append(item)
            continue
        try:
            item["size_kb"] = round(path.stat().st_size / 1024, 1)
        except OSError as e:
            item["size_kb"] = None
            item["error"] = f"读文件大小失败：{e}"
            files.append(item)
            continue
        try:
            reader = PdfReader(str(path))
            n = len(reader.pages)
            item["pages"] = n
            total += n
        except (PdfReadError, ValueError, OSError) as e:
            item["error"] = f"解析失败：{type(e).__name__}: {e}"
        files.append(item)
    return {"files": files, "total_pages": total}

File: mcp/src/tools/pdfTools.ts
/**
 * pdf_tools.* —— PDF 合并 / 拆分 / 检视（C# sidecar，PDFsharp 原子写）。
 *
 * 入参 schema 跟 dotnet/civ-doc/Handlers/PdfToolsHandlers.cs 一一对照。
 */

import { z } from "zod";
import type { ToolDef } from "./registry.js";

export const pdfToolsMerge: ToolDef = {
  rpcMethod: "pdf_tools.merge",
  mcpName: "pdf_tools_merge",
  description:
    "按 inputs 顺序合并多个 PDF 成一个。返回 {output, count}。" +
    "原子写——中途失败不会留半截文件。",
  inputSchema: {
    inputs: z
      .array(z.string())
      .describe("待合并的 PDF 绝对路径数组（按数组顺序拼接）"),
    output: z.string().describe("输出 PDF 绝对路径（所在目录须已存在）"),
  },
};

export const pdfToolsSplitPerPage: ToolDef = {
  rpcMethod: "pdf_tools.split_per_page",
  mcpName: "pdf_tools_split_per_page",
  description:
    "把一个 PDF 按页拆成多个单页 PDF。返回 {written: [path], count}。",
  inputSchema: {
    input: z.string().describe("源 PDF 绝对路径"),
    output_dir: z.string().describe("输出目录绝对路径"),
    name_template: z
      .string()
      .optional()
      .describe(
        "文件名模板，缺省 '{stem}_p{n}.pdf'（{stem}=源文件名去扩展，{n}=零填充页码）",
      ),
  },
};

export const pdfToolsSplitByRanges: ToolDef = {
  rpcMethod: "pdf_tools.split_by_ranges",
  mcpName: "pdf_tools_split_by_ranges",
  description:
    "把一个 PDF 按页码范围表达式拆分。返回 {written: [path], count}。",
  inputSchema: {
    input: z.string().describe("源 PDF 绝对路径"),
    output_dir: z.string().describe("输出目录绝对路径"),
    expr: z
      .string()
      .describe("范围表达式，逗号分隔、连字符表区间，如 '1-3,5,7-9'（1-based）"),
    name_template: z
      .string()
      .optional()
      .describe(
        "文件名模板，缺省 '{stem}_{start}-{end}.pdf'（{start}/{end}=该段起止页）",
      ),
  },
};

export const pdfToolsInspect: ToolDef = {
  rpcMethod: "pdf_tools.inspect",
  mcpName: "pdf_tools_inspect",
  description:
    "检视一批 PDF 的页数和大小。返回 {files: [{path, size_kb, pages?, error?}], total_pages}。" +
    "单个文件失败（不存在/损坏）只在该项标 error，不影响整体。",
  inputSchema: {
    paths: z.array(z.string()).describe("待检视的 PDF 绝对路径数组"),
  },
};

export const allPdfToolsTools: readonly ToolDef[] = [
  pdfToolsMerge,
  pdfToolsSplitPerPage,
  pdfToolsSplitByRanges,
  pdfToolsInspect,
];
